### DISCUSSION TITLE
Formula for subversion now fails to install from source on non default prefix

### DISCUSSION THREAD

**Discussion by Yann-R** [Homebrew/discussions#3123] [Writing Formulae/Casks] [answered]
Hello,

I reported an issue (https://github.com/Homebrew/homebrew-core/issues/93294) but a non default prefix as mine (`/opt/brew`) is not supported.
So I tried to find the bug in the current brew formula (I had no problem with previous versions, it just started to fail to upgrade after version `1.14.1_2`)

The problem seems to be related to the `libserf` (supplied with the formula) and maybe `rpath`, but I really don't manage to correct it.
Any idea to help me?

# Details of the bug report

% brew config
HOMEBREW_VERSION: 3.3.11-2-g02c303c
ORIGIN: https://github.com/Homebrew/brew
HEAD: 02c303c78eec4c4c59174b45edee780fb08249e9
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e4af8b9958c64e14d484956b1fc791fe71b53f19
Core tap last commit: 42 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/brew
HOMEBREW_REPOSITORY: /opt/brew/Homebrew
HOMEBREW_CELLAR: /opt/brew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.c9eHFs6Q4e/org.xquartz:0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /opt/brew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 12.0.0 build 1200
Git: 2.34.1 => /opt/brew/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: 12.4

% brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed with `brew diy`.
You should find replacements for the following formulae:
  xtail
  colorsvn
  contacts
  godep
  csv-fix

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  archivemount
  avfs
  curlftpfs
  ext2fuse
  ext4fuse
  fuse-zip
  godep
  growly
  ifuse
  isl@0.18
  mp3fs
  ntfs-3g
  q
  rdesktop
  sdl
  sdl_image
  sdl_mixer
  sdl_ttf
  sshfs
  xmount

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/ublio.h

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  unison

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.

# Verification

- I ran brew update and am still able to reproduce my issue.
- I have resolved all warnings from brew doctor and that did not fix my problem.

# What were you trying to do (and why)?

I typed:
brew upgrade subversion

To upgrade after brew outdated mentioned this:
subversion (1.14.1_2) < 1.14.1_4

# What happened (include all command output)?

```
% brew upgrade subversion
==> Upgrading 1 outdated package:
subversion 1.14.1_2 -> 1.14.1_4
Warning: Building subversion from source as the bottle needs:
 HOMEBREW_CELLAR: /usr/local/Cellar (yours is /opt/brew/Cellar)
 HOMEBREW_PREFIX: /usr/local (yours is /opt/brew)
==> Downloading https://raw.githubusercontent.com/Homebrew/formula-patches/03cf8088210822aa2c1ab544ed58ea04c897d9c4/libtool/configure-big_sur.diff
Already downloaded: /Users/ricquebo/Library/Caches/Homebrew/downloads/195208ba31cadc265ea97bb8aad4e78caab1200daf4d855e6c839b5f987f92db--configure-big_sur.diff
==> Downloading https://github.com/encukou/py3c/archive/v1.1.tar.gz
Already downloaded: /Users/ricquebo/Library/Caches/Homebrew/downloads/341faafd274c164961d080ffbaab85a010d9385563dc574212244e60297de7b8--py3c-1.1.tar.gz
==> Downloading https://www.apache.org/dyn/closer.lua?path=serf/serf-1.3.9.tar.bz2
Already downloaded: /Users/ricquebo/Library/Caches/Homebrew/downloads/cfe7b56061b1bb57199b8b7a8782c47f1e318dbf2f6e108ce67c20f926be1df9--serf-1.3.9.tar.bz2
==> Downloading https://www.apache.org/dyn/closer.lua?path=subversion/subversion-1.14.1.tar.bz2
Already downloaded: /Users/ricquebo/Library/Caches/Homebrew/downloads/ccce2e737799963f7ff80a0a71b827faefc77379679e596751fa652a61d4be38--subversion-1.14.1.tar.bz2
==> Upgrading subversion
  1.14.1_2 -> 1.14.1_4 

==> Patching
==> Applying configure-big_sur.diff
patching file configure
Hunk #1 succeeded at 10798 (offset 2065 lines).
==> scons PREFIX=/opt/brew/Cellar/subversion/1.14.1_4/libexec/serf GSSAPI=/usr CC=/usr/bin/clang CFLAGS=-Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/Comman
Last 15 lines from /Users/ricquebo/Library/Logs/Homebrew/subversion/01.scons:
/usr/bin/clang -o buckets/response_buckets.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include buckets/response_buckets.c
/usr/bin/clang -o buckets/simple_buckets.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include buckets/simple_buckets.c
/usr/bin/clang -o buckets/socket_buckets.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include buckets/socket_buckets.c
/usr/bin/clang -o buckets/ssl_buckets.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include buckets/ssl_buckets.c
/usr/bin/clang -o auth/auth.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth.c
/usr/bin/clang -o auth/auth_basic.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth_basic.c
/usr/bin/clang -o auth/auth_digest.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth_digest.c
/usr/bin/clang -o auth/auth_spnego.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth_spnego.c
/usr/bin/clang -o auth/auth_spnego_gss.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth_spnego_gss.c
/usr/bin/clang -o auth/auth_spnego_sspi.os -c -Os -w -pipe -march=nehalem -mmacosx-version-min=11 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -dynamic -fPIC -DNDEBUG -DDARWIN -DSIGPROCMASK_SETS_THREAD_MASK -DDARWIN_10 -DSERF_HAVE_GSSAPI -I. -I/opt/brew/opt/apr/libexec/include/apr-1 -I/opt/brew/opt/apr-util/libexec/include/apr-1 -I/opt/brew/opt/openssl@1.1/include -I/opt/brew/opt/openssl@1.1/include auth/auth_spnego_sspi.c
/usr/bin/clang -o libserf-1.1.3.0.dylib -L/opt/brew/opt/openssl@1.1/lib -L/opt/brew/opt/apr/lib -L/opt/brew/opt/python@3.10/lib -L/opt/brew/lib -F/opt/brew/Frameworks -Wl,-headerpad_max_install_names -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -Wl,-install_name,/private/tmp/subversion--serf-20220329-16251-alihpq/serf-1.3.9/libserf-1.dylib -Wl,-search_paths_first -dynamiclib -Wl,-current_version,1.3.0 -Wl,-compatibility_version,1.3.0 -Wl,-rpath=/opt/brew/Cellar/subversion/1.14.1_4/libexec/serf/lib -Wl,-rpath=/opt/brew/opt/apr/libexec/lib -Wl,-rpath=/opt/brew/opt/openssl@1.1/lib -Wl,-rpath=/opt/brew/opt/apr-util/libexec/lib -Wl,-rpath=/opt/brew/opt/openssl@1.1/lib context.os incoming.os outgoing.os ssltunnel.os buckets/aggregate_buckets.os buckets/allocator.os buckets/barrier_buckets.os buckets/buckets.os buckets/bwtp_buckets.os buckets/chunk_buckets.os buckets/dechunk_buckets.os buckets/deflate_buckets.os buckets/file_buckets.os buckets/headers_buckets.os buckets/iovec_buckets.os buckets/limit_buckets.os buckets/mmap_buckets.os buckets/request_buckets.os buckets/response_body_buckets.os buckets/response_buckets.os buckets/simple_buckets.os buckets/socket_buckets.os buckets/ssl_buckets.os auth/auth.os auth/auth_basic.os auth/auth_digest.os auth/auth_spnego.os auth/auth_spnego_gss.os auth/auth_spnego_sspi.os -L/opt/brew/opt/apr/libexec/lib -L/opt/brew/opt/openssl@1.1/lib -L/opt/brew/opt/apr-util/libexec/lib -L/opt/brew/opt/openssl@1.1/lib -lssl -lcrypto -lz -lapr-1 -lpthread -laprutil-1 -lexpat -liconv -lkrb5
ld: unknown option: -rpath=/opt/brew/Cellar/subversion/1.14.1_4/libexec/serf/lib
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [libserf-1.1.3.0.dylib] Error 1
scons: building terminated because of errors.
```

# What did you expect to happen?

Installation of subversion 1.14.1_4 finishing without failing.

# Step-by-step reproduction instructions (by running brew commands)
brew upgrade subversion

**Comment by xlla** [comment]
same issue here on subversion 1.14.1_4

<pre>
/usr/bin/clang -o libserf-1.1.3.0.dylib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/apr/lib -L/usr/local/opt/python@3.10/lib -L/usr/local/lib -F/usr/local/Frameworks -Wl,-headerpad_max_install_names -Wl,-install_name,/private/tmp/subversion--serf-20220401-55694-tcm61h/serf-1.3.9/libserf-1.dylib -Wl,-search_paths_first -dynamiclib -Wl,-current_version,1.3.0 -Wl,-compatibility_version,1.3.0 -Wl,-rpath=/usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib -Wl,-rpath=/usr/local/opt/apr/lib -Wl,-rpath=/usr/local/opt/openssl@1.1/lib -Wl,-rpath=/usr/local/opt/apr-util/libexec/lib -Wl,-rpath=/usr/local/opt/openssl@1.1/lib context.os incoming.os outgoing.os ssltunnel.os buckets/aggregate_buckets.os buckets/allocator.os buckets/barrier_buckets.os buckets/buckets.os buckets/bwtp_buckets.os buckets/chunk_buckets.os buckets/dechunk_buckets.os buckets/deflate_buckets.os buckets/file_buckets.os buckets/headers_buckets.os buckets/iovec_buckets.os buckets/limit_buckets.os buckets/mmap_buckets.os buckets/request_buckets.os buckets/response_body_buckets.os buckets/response_buckets.os buckets/simple_buckets.os buckets/socket_buckets.os buckets/ssl_buckets.os auth/auth.os auth/auth_basic.os auth/auth_digest.os auth/auth_spnego.os auth/auth_spnego_gss.os auth/auth_spnego_sspi.os -L/usr/local/opt/apr/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/apr-util/libexec/lib -L/usr/local/opt/openssl@1.1/lib -lssl -lcrypto -lz -lapr-1 -lpthread -laprutil-1 -lexpat -liconv -lkrb5
ld: unknown option: -rpath=/usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [libserf-1.1.3.0.dylib] Error 1
scons: building terminated because of errors.

</pre>

**Comment by Yann-R** [comment]
Nice to join the discussion.
@cabo who contributed in the issue mentioned above seems also to have faced the same error.

**Comment by cabo** [comment]
So the problem is that clang's ld expects

```
-rpath /usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib
```

and gets

``` 
-rpath=/usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib
```

which is because clang got passed

``` 
-Wl,-rpath=/usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib
``` 

and not

``` 
-Wl,-rpath,/usr/local/Cellar/subversion/1.14.1_4/libexec/serf/lib
```

Note that the form with the comma is portable, while the form with the equals sign doesn't work with clang.

I'm not enough of a scons archeologist to find the root cause, but it should be trivial.

**Reply by Bo98** [reply]
That's correct, but scons should already handle this and it does for me. I'm not able to reproduce the issue with `brew install -s subversion` so I wonder what's different that's causing scons to not correctly detect you are on macOS.

**Comment by Yann-R** [comment]
@cabo 
> Note that the form with the comma is portable, while the form with the equals sign doesn't work with clang.

Thanks a lot for that great progress for me: you pointed out the source of the problem. 👍

@Bo98 
> scons should already handle this and it does for me. I'm not able to reproduce the issue with brew install -s subversion so I wonder what's different that's causing scons to not correctly detect you are on macOS.

Could the context of the bug be "more tricky", I mean:
- are you using macOS?
- are you using (like me) a non standard homebrew prefix (/opt/brew) for an Intel-based mac?
Maybe scons behaves perflecty to handle macOS when the things are not different?

And wouldn't there be some "magic incantations" (with env. variables?) that scons gurus could know to "force" the determination of macOS platform?

**Comment by Yann-R** [comment]
Hi all,

I observed that the current formula for  [`subversion`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/subversion.rb) supports macOS & Linux, and has (strangely?) a special part for `libserf` on Linux:
```
    if OS.linux?
      # svn can't find libserf-1.so.1 at runtime without this
      ENV.append "LDFLAGS", "-Wl,-rpath=#{serf_prefix}/lib"
    end
```

I tried to play a little with this for macOS (by removing the condition and adapting the syntax for `rpart` as mentionned by @cabo to remove the `=` and use `,`).
But it has no effect: in fact my modified LDFLAGS do not appear anywhere at all in the failing compilation command appearing in my first message.
No more ideas...

**Reply by Bo98** [reply]
We added a tweak recently with 1.14.2 that was supposed to cover the issue: https://github.com/Homebrew/homebrew-core/blob/4eff6eb4492e4658f17c8d4af5b0ffeee1f20b69/Formula/subversion.rb#L100-L101

Did it not work for you?

**Reply by Yann-R** [reply]
Oh yes, thanks a lot @Bo98, I was not yet aware of this recent tweak.
`subversion`was pinned on my system to avoid failure at each upgrade of all my brew formulae.

This solves my problem! Best regards.

**Comment by Yann-R** [comment, marked as answer]
Solved by the tweak mentioned above.
